### PR TITLE
two-pass label shader

### PR DESCRIPTION
**Pull request by hjanetzek**
Draw fill and stroke in separate passes to prevent drawing strokes over other glyphs fill.

Should resolve issue https://github.com/tangrams/tangram-es/issues/460

### CODE REVIEW

**Comment by karimnaaji** [comment]
This fixes the issue, we can ask for a two pass only when the style has stroke params on.

File: core/resources/shaders/sdf.fs [hunk 1]
```diff
@@ -68,16 +69,20 @@ float sampleAlpha(in vec2 uv, float distance, float threshold) {
...
-
-    float alpha_fill = pow(sampleAlpha(v_texcoords, distance, threshold_fill), 0.4545);
-    float alpha_stroke = pow(sampleAlpha(v_texcoords, distance, threshold_stroke), 0.4545);
-
-    vec4 color = mix(v_strokeColor, v_color, alpha_fill);
-    color.a = max(alpha_fill, alpha_stroke) * v_alpha * v_color.a;
+    vec4 color;
+
+    if (u_pass == 1) {
+      float threshold_fill = 0.5;
```
Comment (karimnaaji) [inline core/resources/shaders/sdf.fs]: nitpick: can we keep the 4 spaces tabs.

**Comment by hjanetzek** [comment]
@karimnaaji do you mean to check whether any of the added labels has stroke?

**Comment by karimnaaji** [comment]
Yeah exactly what's done with https://github.com/tangrams/tangram-es/commit/0e9e57ad2824cd43d33f14b61316c054947d5c9d

File: core/resources/shaders/sdf.fs [hunk 2]
```diff
@@ -68,16 +69,20 @@ float sampleAlpha(in vec2 uv, float distance, float threshold) {
...
     float distance = texture2D(u_tex, v_texcoords).a;
-
-    float alpha_fill = pow(sampleAlpha(v_texcoords, distance, threshold_fill), 0.4545);
-    float alpha_stroke = pow(sampleAlpha(v_texcoords, distance, threshold_stroke), 0.4545);
-
-    vec4 color = mix(v_strokeColor, v_color, alpha_fill);
-    color.a = max(alpha_fill, alpha_stroke) * v_alpha * v_color.a;
+    vec4 color;
+
+    if (u_pass == 1) {
```
Comment (bcamper) [inline core/resources/shaders/sdf.fs]: I know we're not filling a lot of fragments with SDF shaders, but do we know the trade-off in a branching shader vs. compiling a separate shader for each pass?

Comment (hjanetzek) [inline core/resources/shaders/sdf.fs]: Branching on uniforms is quite cheap as far as I know, as it should be easy to predict which branch to take. Both branches must only be computed when both can be reached within a shader unit (those that are processing e.g. 16x16 pixel in parallel with one process counter). That's just guessing though, but may the reason why apple states that branching on uniforms is acceptable: https://developer.apple.com/library/ios/documentation/3DDrawing/Conceptual/OpenGLES_ProgrammingGuide/BestPracticesforShaders/BestPracticesforShaders.html

I'll see if it is possible to move the branching to the vertex shader

Comment (bcamper) [inline core/resources/shaders/sdf.fs]: Thanks, just curious.

**Comment by blair1618** [comment]
Nice, that was :zap: fast.

One very very minor optimization if it makes sense to you: We can prevent text without stroke from changing any uniforms. Let `u_pass == 0` correspond to the fill pass, then if `m_strokePass` we can set `u_pass` to 1, draw, then reset it to 0. 

Draw calls on text buffers are actually quite an expensive part of our scenes right now, presumably because the buffers contain all occluded labels as well as the visible ones. Whatever the cause, text draw calls consume upwards of 70% of render time on low-end devices, so optimizations here are not premature.

**Comment by hjanetzek** [comment]
@blair1618, yes, the uniform optimization makes sense.

The number of discarded labels that are sent with each draw call are also my concern. Generally I would propose to use a single mesh (possibly double buffered) to collect all active label vertices. The updates could also be done first in a render-loop iteration so that upload happens during static meshes are drawn.

**Comment by blair1618** [comment]
Cool, this is looking good to me. I'll merge in an hour if there are no further comments.

**Pull request merged**